dipy/align/imaffine.py

```diff
@@ -0,0 +1,1010 @@
+                                            grid_to_world)
+                # Dense case: we just need to provide the grid-to-world
+                # transform to obtain the sampling points' world coordinates
+                # because we know that all grid cells must be processed
+                sampling_info = grid_to_world
+            else:
+                # Compute the gradient of moving at the sampling points
+                # which are already given in physical space coordinates
+                moving32 = self.moving.astype(np.float32)
+                mgrad, inside = vf.sparse_gradient(moving32,
+                                                   sp_to_moving,
+                                                   self.moving_spacing,
+                                                   self.samples)
+                # Sparse case: we need to provide the actual coordinates of
+                # all sampling points
+                sampling_info = self.samples[..., :self.dim]
+            self.update_gradient(params, self.transform, static_values,
+                                 moving_values, sampling_info, mgrad)
+
+        # Evaluate the mutual information and its gradient
+        # The results are in self.metric_val and self.metric_grad
+        # ready to be returned from 'distance' and 'gradient'
+        self.update_mi_metric(update_gradient)
+        return True
+
+    def distance(self, params):
+        r""" Numeric value of the negative Mutual Information
+
+        We need to change the sign so we can use standard minimization
+        algorithms.
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        neg_mi : float
+            the negative mutual information of the input images after
+            transforming the moving image by the currently set transform
+            with `params` parameters
+        """
+        if not self._update(params, False):
+            return np.inf
+        return -1 * self.metric_val
+
+    def gradient(self, params):
+        r""" Numeric value of the metric's gradient at the given parameters
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        grad : array, shape (n,)
+            the gradient of the negative Mutual Information
+        """
+        if not self._update(params, True):
+            return self.metric_grad
+        return -1 * self.metric_grad
+
+    def distance_and_gradient(self, params):
```

Comment:
This is just to comply with the optimization API: we can either provide the Optimizer with the metric and gradient separately, or we can return both from a single function. I don't think this is ugly, it makes sense because here:
https://github.com/omarocegueda/dipy/blob/imaffine_pr3/dipy/align/parzenhist.pyx#L1286

 the metric and gradient can be computed within a single nested loop (cheaper than making two separate calls, one for metric and one for gradient) and we can save time by not executing the gradient loop when it's not necessary.